### PR TITLE
added deterministic opinfo attribute

### PR DESCRIPTION
**Pull request by srossross**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85101
* #85011

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -194,7 +194,11 @@ def to_cpu(arg):
             self.assertEqual(len(cuda_results), len(cpu_results))
 
             for cpu_result, cuda_result in zip(cpu_results, cuda_results):
-                self.assertEqual(cpu_result, cuda_result)
+                if not op.deterministic and isinstance(cpu_result, torch.Tensor):
+                    self.assertEqual(cpu_result.dtype, cuda_result.dtype)
+                    self.assertEqual(cpu_result.shape, cuda_result.shape)
```
Comment (srossross) [inline test/test_ops.py:199]: @mruberry @amjames here is what I was thinking for an `op. deterministic` flag. In this context the asserts above like `self.assertEqual(len(cuda_results), len(cpu_results))` make more sense. 

We can do a more fine grained check for both `deterministic=True` and `deterministic=False`

Comment (amjames) [reply test/test_ops.py:200]: Shape/dtype checks seem sufficient.

File: test/test_ops.py [hunk 2]
```diff
@@ -194,7 +195,11 @@ def to_cpu(arg):
             self.assertEqual(len(cuda_results), len(cpu_results))
 
             for cpu_result, cuda_result in zip(cpu_results, cuda_results):
-                self.assertEqual(cpu_result, cuda_result)
+                if not op.deterministic and isinstance(cpu_result, torch.Tensor):
```
Comment (amjames) [inline test/test_ops.py:198]: If the op is non deterministic and the result is not a tensor the else condition will catch. We probably don't want to do equality comparison on the results of non-deterministic ops, I can't think of any examples off the top of my head that would trigger this but better to be precise.

Comment (srossross) [reply test/test_ops.py:198]: were you thinking of something like:

```py
                if op.deterministic:
                    self.assertEqual(cpu_result, cuda_result)
                elif isinstance(cpu_result, torch.Tensor):
                    self.assertEqual(cpu_result.dtype, cuda_result.dtype)
                    self.assertEqual(cpu_result.shape, cuda_result.shape)
                else:
                    # This is not a current use-case, all non-deterministic ops return a tensor
                    pass
```

**Comment by mruberry** [comment]
I'd be OK with adding deterministic metadata to operators (cc @kurtamohler), but I'm not sure now is the time to do so. What would be nice is if we were using OpInfos for metadata outside testing, and so could use that metadata to do things like implement the nondeterminism check, but that's not where we are today

**Comment by srossross** [comment]
closing this and will revisit when appropriate.

**Pull request closed without merging**